subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/mcm/analysis/FiniteStateChecker.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright 2023 Budapest University of Technology and Economics
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package hu.bme.mit.theta.analysis.algorithm.mcm.analysis
+
+import hu.bme.mit.theta.analysis.InitFunc
+import hu.bme.mit.theta.analysis.LTS
+import hu.bme.mit.theta.analysis.TransFunc
+import hu.bme.mit.theta.analysis.algorithm.SafetyChecker
+import hu.bme.mit.theta.analysis.algorithm.SafetyResult
+import hu.bme.mit.theta.analysis.algorithm.mcm.interpreter.MemoryEventProvider
+import hu.bme.mit.theta.analysis.expl.ExplPrec
+import hu.bme.mit.theta.analysis.expl.ExplState
+import hu.bme.mit.theta.analysis.expr.StmtAction
+import hu.bme.mit.theta.common.Tuple
+import hu.bme.mit.theta.common.Tuple1
+import hu.bme.mit.theta.common.Tuple2
+import hu.bme.mit.theta.graphsolver.compilers.GraphPatternCompiler
+import hu.bme.mit.theta.graphsolver.patterns.constraints.GraphConstraint
+import hu.bme.mit.theta.graphsolver.solvers.GraphSolver
+import java.util.*
+
+class FiniteStateChecker<S : ExplState, A : StmtAction, T>(
```

Comment:
Please add documentation (or links to papers) that describe what this algorithm can be used for.